### ISSUE TITLE
Wrong narrowing with the `in` operator

### ISSUE THREAD

**Issue by AxelHaddad** [microsoft/TypeScript#53910]
# Bug Report

It seems that the [built-in narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) via the `in` operator  can lead to narrowing mistakes (false positive).

### 🔎 Search Terms

Narrowing, `in`, subtypes, type guards

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about type guards/narrowing

### ⏯ Playground Link

- a first [Playground](https://www.typescriptlang.org/play?ts=5.0.4&ssl=11&ssc=2&pln=4&pc=1#code/PTAEHUAsE9QVwM4EsB2BzUAXSBTUADVfUAax2gHcB7AJwBNQBjGnAQ0xwVADlWaaqFVBhz9aAKEmZoABzwAFPpiSsANgBVZeALygA3gDMqVAFygEmGsIC+46XNAAxOKo1bQuw8ZMWr6ADSgAEZ8PpbCgSEAXma+NpIGcCiMylQooEgIzq6acgAUVEEAVjgpZtluDgA+oIo0ymq5OACUZoUlKRlcFU364qADGQageQCEeQBEITQTGentpZjNzaAsmHA06QZqCDgA3P2DjGkWoAD6Jj3uugspe6AgoABSiJhYVKBCmCicXNjsoCaCGYSBkb2wqBIXFub0yThclRwhwGaw26UscH24ls4mOKFOAFtoAB5YqLDz6IymCa4VxUCaRUKgCbUGiqOgTWx4wnQOoNRF0UkdTBmPkqREUolCxbiJAGPKZK75IlixpaQVklLLPTIpgnKiqHAAOlUVDQkwA4lRMJhGJBWBNmtjxEA) leading to false narrowing
- A second [Playground](https://www.typescriptlang.org/play?ts=5.0.4#code/PTAEHUAsE9QVwM4EsB2BzUAXSBTUADVfUAax2gHcB7AJwBNQBjGnAQ0xwVADlWaaqFVBhz9aAKEmZoABzwAFPpiSsANgBVZeALygA3gDMqVAFygEmGsIC+46XNAAxOKo1bQuw8ZMWr6ADSgAEZ8PpbCgSEAXmZ6QaqsZr42tuIGcCiMylQooAC20M6ZABRUQQBWOFlmzq6aDgA+oIo0ymr1OACUZgBuVEgMeuKgI6BIBqDFAITFAEQhNLNjuWWVWZ2d+sOjO4w5CFSqOAB0qlRocyhUmE4ubnKznQDc2zsjLJhwNCivoLa7+0OJzOF1mtXuOEevz2KAsoAA+iZwR0PKBVlVME9QCBQAApRA3TBUUBCTAoThcbDsUAdBDMJAyQmQVAkLjorJjLjIrTQwFHU7nUoVDHHaKihKdcSpGFwgoAeWFHM8RlMs1wrios0ioVAs2oNFUdFmthlNwKLTaELoCrWmDMFpUENR8sVmHEBSKjGK5qUjo61tdkqAA) leading to uncatched type errors that breaks at runtime. 

### 💻 Code

```ts
type PartialType = {foo: string}
type FullType = {foo:string, bar:string, baz: string}

function isFullType(object: FullType | PartialType): object is FullType {
    if (!("bar" in object)) return false;
    const _:FullType = object; // Just to witness that Typescript thinks object is FullType
    return true;
}
```

### 🙁 Actual behavior

No typescript error.

### 🙂 Expected behavior

Typescript should not think that  `object` is of type `FullType` as for instance all those three following objects, compatible with `PartialType`, are clearly not `FullType`:
```ts
{ foo: "hello", bar: "world" }
{ foo: "hello", bar: undefined, baz: 24 }
{ foo: "hello", bar: 42 }
```
Therefore it should raise an error at the line  `const _:FullType = object;`.

**Comment by MartinJohns** [comment]
Duplicate of #34975. The `in` operator is intentionally unsound.

**Comment by ziloen** [comment]
Maybe you should use `type PartialType = { foo: string } & { [K in string]: unknown }`

**Comment by AxelHaddad** [comment]
Ok, I understand that if I were to use the `in` keyword, I accept that Typescript won't be sound anymore.

That being said, how can I , without using `in`, correctly narrow an object `obj` of type `PartialType = {foo: string}` to the  type `FullType = {foo:string, bar:string, baz: string}` since I cannot call `obj.bar` or `obj["bar"]`  without asserting `"bar" in obj` first?

**Comment by guillaumebrunerie** [comment]
@AxelHaddad Use discriminated unions instead, it is much more sound than regular unions. That is use the following two types:
```typescript
type PartialType = {isFull: false, foo: string}
type FullType = {isFull: true, foo:string, bar:string, baz: string}
```
and check the value of `isFull`.
